x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,431 @@
+	c := make(chan HealthDetails, maxResults)
+
+	for {
+		// When invoking the DescribeEvents for the first time, there must not exist any NextToken.
+		// Upon calling the DescribeEvents API, the next token will be returned if there are additional records available for querying.
+		// If there are no more records to fetch, the next token will be empty.
+		if nextTokenString == "" {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+				},
+			)
+		} else {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+					NextToken:  &nextTokenString,
+				},
+			)
+		}
+
+		if err != nil {
+			m.Logger().Errorf("[AWS Health] DescribeEvents failed with : %w", err)
+			return nil
+		}
+		ets := eventOutput.Events
+		select {
+		case <-ctx.Done():
+			// Context cancelled, handle graceful termination
+			close(c)
+			return nil
+		default:
+			// Context not cancelled, proceed with the function
+		}
+
+		for i := range ets {
+			m.Logger().Debugf("[AWS Health] [Fetch DescribeEventDetails] Event ARN : %s", getValueOrDefault(ets[i].Arn, ""))
+			go func(et types.Event) {
+				err := m.getDescribeEventDetails(ctx, awsHealth, et, c)
+				if err != nil {
+					errCh <- err
+				}
+			}(ets[i])
+		}
+
+		for i := 0; i < len(ets); i++ {
+			select {
+			case <-ctx.Done():
+				// Context cancelled, handle graceful termination
+				m.Logger().Debug("Context cancelled. Exiting gracefully.")
+				close(c)
+				return nil
+			case err := <-errCh:
+				// Handle errors received from goroutines
+				m.Logger().Error(err.Error())
+			case healthDetails, ok := <-c:
+				if !ok {
+					return nil
+				}
+				m.Logger().Debugf("[AWS Health] [DescribeEventDetails] Event ARN : %s, Affected Entities (Pending) : %d, Affected Entities (Resolved): %d, Affected Entities (Others) : %d", *healthDetails.event.Arn, healthDetails.affectedEntityPending, healthDetails.affectedEntityResolved, healthDetails.affectedEntityOthers)
+				events = append(events, createEvents(healthDetails))
+			}
+		}
+		if eventOutput.NextToken == nil {
+			break
+		}
+		nextTokenString = *eventOutput.NextToken
+	}
+	close(c)
+	close(errCh)
+	return events
+}
+
+// createEvents takes in a HealthDetails struct and returns an mb.Event struct
+// populated with the data from the HealthDetails. It sets the MetricSetFields,
+// RootFields and ID fields of the mb.Event struct. The MetricSetFields contain
+// the details of the health event. The RootFields specify that it is an AWS
+// event. The ID is generated from the event ARN, status code and current date.
+func createEvents(hd HealthDetails) mb.Event {
+	currentDate := getCurrentDateTime()
+	eventID := currentDate + getValueOrDefault(hd.event.Arn, "") + getValueOrDefault((*string)(&hd.event.StatusCode), "")
```

Comment:
we should use the [built in pointer conversion utilies](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws#hdr-Value_and_Pointer_Conversion_Utilities) for safe dereferencing